sdks/java/io/parquet/src/main/java/org/apache/beam/sdk/io/parquet/ParquetIO.java

```diff
@@ -230,14 +279,255 @@ public ReadFiles withAvroDataModel(GenericData model) {
+          out.output(offsetRange);
+        }
+      }
+
+      public ArrayList<OffsetRange> splitBlockWithLimit(
+          long start, long end, List<BlockMetaData> blockList, long limit) {
+        ArrayList<OffsetRange> offsetList = new ArrayList<OffsetRange>();
+        long totalSize = 0;
+        long rangeStart = start;
+        long rangeEnd = start;
+        for (long i = start; i < end; i++) {
+          totalSize += blockList.get((int) i).getTotalByteSize();
+          rangeEnd += 1;
+          if (totalSize >= limit) {
+            offsetList.add(new OffsetRange(rangeStart, rangeEnd));
+            rangeStart = rangeEnd;
+            totalSize = 0;
+          }
+        }
+        if (totalSize != 0) {
+          offsetList.add(new OffsetRange(rangeStart, rangeEnd));
+        }
+        return offsetList;
+      }
+
+      @NewTracker
+      public RestrictionTracker<OffsetRange, Long> newTracker(
+          @Restriction OffsetRange restriction, @Element FileIO.ReadableFile file)
+          throws Exception {
+        List<Double> recordCountAndSize = getRecordCountAndSize(file, restriction);
+        return new BlockTracker(
+            restriction,
+            Math.round(recordCountAndSize.get(1)),
+            Math.round(recordCountAndSize.get(0)));
+      }
+
+      @GetRestrictionCoder
+      public OffsetRange.Coder getRestrictionCoder() {
+        return new OffsetRange.Coder();
+      }
+
+      @GetSize
+      public double getSize(@Element FileIO.ReadableFile file, @Restriction OffsetRange restriction)
+          throws Exception {
+        return getRecordCountAndSize(file, restriction).get(1);
+      }
+
+      public List<Double> getRecordCountAndSize(
+          @Element FileIO.ReadableFile file, @Restriction OffsetRange restriction)
+          throws Exception {
+        ParquetFileReader reader = getParquetFileReader(file);
+        double size = 0;
+        double recordCount = 0;
+        for (long i = restriction.getFrom(); i < restriction.getTo(); i++) {
+          BlockMetaData block = reader.getRowGroups().get((int) i);
+          recordCount += block.getRowCount();
+          size += block.getTotalByteSize();
+        }
+        List<Double> countAndSize = new ArrayList<>();
+        countAndSize.add(recordCount);
+        countAndSize.add(size);
+        return countAndSize;
+      }
+    }
+
+    public static class BlockTracker extends OffsetRangeTracker {
+      private long totalWork;
+      private long progress;
+      private long approximateRecordSize;
+
+      public BlockTracker(OffsetRange range, long totalByteSize, long recordCount) {
+        super(range);
+        if (recordCount != 0) {
+          this.approximateRecordSize = totalByteSize / recordCount;
+          this.totalWork = approximateRecordSize * recordCount;
+          this.progress = 0;
+        }
+      }
+
+      public void makeProgress() throws Exception {
+        progress += approximateRecordSize;
```

Comment:
It seems like in the DoFn body, you will call `tracker.makeProgress()` for each record. It doesn't work under fnapi context since we only expose limit APIs: https://github.com/apache/beam/blob/7fb07ff78e9895fb486f25f2ee477b875b112c3b/sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/splittabledofn/RestrictionTrackers.java#L91-L104

Reply:
Do you mean that we do not pass in the restriction tracker created by the user to the DoFn but instead pass in a delegate ? If so, I think this is a significant limitation since that will not allow users to adequately customize the way RestrictionTracker is used.

Reply:
Yes, we wrap the user-defined tracker with certain locking guarantee.

cc @lukecwik for exposing APIs discussion.

The purpose of wrapper is to reduce user effort on complex things, like locking. And in most case. we don't really expect the user to alter the tracker by calling non-exposing APIs.

Reply:
For any block based source (Avro, Parquet, etc), I think it's important to report an approximate progress within blocks instead of just reporting the progress of blocks. We do this for the current sources: https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/io/BlockBasedSource.java#L241

cc: @lukecwik @robertwb

Reply:
@chamikaramj Customizing the restriction tracker allows the user to mutate it's state in many ways and leads to lots of concurrency issues and also correctness guarantees.

Reply:
I agree that current implementation might be safer but seems like it significantly limits the usability of RestrictionTrackers as well. I think source authors are advanced users. So it might be OK to stay in the side of usability in some cases over safety. But this is debatable.

Another option might be to add a method to the RestrictionTracker to mark progress without claiming. We have such a method in the old RangeTracker interface: https://github.com/apache/beam/blob/master/sdks/python/apache_beam/io/iobase.py#L393